R/shiny.R

```diff
@@ -962,20 +962,37 @@ ShinySession <- R6Class(
     },
     flushOutput = function() {
 
+      # Return TRUE if there's any stuff to send to the client.
+      hasPendingUpdates <- function() {
+        return(
+          length(private$invalidatedOutputValues) != 0 ||
+          length(private$invalidatedOutputErrors) != 0 ||
+          length(private$inputMessageQueue) != 0
+        )
+      }
+
       # ..stacktraceon matches with the top-level ..stacktraceoff..
       private$flushCallbacks$invoke(..stacktraceon = TRUE)
-      # ..stacktraceon matches with the top-level ..stacktraceoff..
-      on.exit(private$flushedCallbacks$invoke(..stacktraceon = TRUE))
 
-      if (length(private$progressKeys) == 0
```

Comment:
The original code was correct. The existence of progress keys means we need to send a `values` message (even if empty) as this has a side effect of clearing out progress indication on the client. See the other note I added.